exploration/bidi-usability.md

```diff
@@ -431,7 +433,17 @@ Add isolates and strongly directional marks to required and optional whitespace
 This would permit users to get the effects described by the above design,
 as long as they use isolates/marks in a "responsible" way.
 
-(Omitting other changes found in #673)
+The exception to this is the namespace separator, used in `identifier`.
+This requires the ability to insert isolates or strongly directional marks
+between the namespace and name portions, where whitespace is not permitted.
+This is the only location in the syntax where such characters might be needed
+but whitespace is not at least optional.
+This could be defined as:
+```abnf
+ns-separator   = [bidi] ":" [bidi]
```

Comment:
Why allow for the `bidi` after `:` if we're not allowing it to show up elsewhere after a starting sigil?

Reply:
Your example `$_مصر‎⁩` can be made to render correctly using LRI/PDI around the identifier (which the syntax allows and the strict form encourages). The `[bidi]` production allows starting or ending strongly directional marks to prevent spillover in the middle of a namespaced identifier, e.g. `$_م1صر:‎_م2صر‎⁩`

I'll admit that it's generally better practice to put the strongly directional character at the end (before the `:` separator), but I didn't want to make the syntax ultra-fussy: whatever stew of strongly directional characters and a colon are not part of either the `namespace` or the `name`.

The separator is different from sigils because the sigils are all at token-start, whereas the namespace separator is embedded into the word-token.

code points for the first example: `\u2066$_\u0645\u0635\u0631\u200e\u2069` [try it](https://w3c.github.io/i18n-discuss/explainers/bidi-html-demo.html?text=\u2066$_\u0645\u0635\u0631\u200e\u2069)
code points for the second example: `\u2066$_\u06451\u0635\u0631:\u200e_\u06452\u0635\u0631\u200e\u2069` [try it](https://w3c.github.io/i18n-discuss/explainers/bidi-html-demo.html?text=\u2066$_\u06451\u0635\u0631:\u200e_\u06452\u0635\u0631\u200e\u2069)

Reply:
> Your example `$_مصر‎⁩` can be made to render correctly using LRI/PDI around the identifier (which the syntax allows and the strict form encourages). [...]
> 
> code points for the first example: `\u2066$_\u0645\u0635\u0631\u200e\u2069` [try it](https://w3c.github.io/i18n-discuss/explainers/bidi-html-demo.html?text=%5Cu2066$_%5Cu0645%5Cu0635%5Cu0631%5Cu200e%5Cu2069)

That's rendering for me with the `_` next to the `$`, which is wrong? The `_` here has neutral directionality, and it's the first character of a name which as a whole is RTL. So we ought to have some way to render it as the right-most character, but that's not possible without some bidi control after the `$`.

Reply:
Fair enough, although a display like `$⁧_مصر⁩⁩` is in itself super-weird (and it's namespaced friend `$⁧_م1صر⁩:⁧_م2صر⁩⁩` is even weirder) from the point of view that it's an RTL isolated run inside an LTR isolated run (or, in the latter case, two RTL runs inside an LTR run). This makes the name tokens display correctly RTL while still forcing the overall token (including the sigil) to display LTR. That is, this sequence:

> \u2066$\u2067_\u06451\u0635\u0631\u2069:\u2067_\u06452\u0635\u0631\u2069\u2069

... with 6 of the 18 code points being bidi isolates. 

There's a similar problem with options, where we're trying to force them to be LTR-ordered (`option = value`/`م1صر‎=م2صر⁩`) when RTL really _really_ wants that to be displayed as ` ⁦م1صر=م2صر⁩`.

So you're right: my proposal does compromise some aspects of RTL display as part of our insistence that messages are intended to work in an LTR editing environment with LTR syntax. Either way, logical order is all that matters and users should be careful about using mixed direction for non-literal values. We give them the tools to make their bidi literals and patterns work RTL-normally as well as the tools to let non-RTL speakers read placeholders LTR (like most developers debugging messages). It is not perfect. Do we need that last bit of cruft to allow the sigil and identifier to be separate? (Not asking rhetorically, btw. What do people think? Where can we get developer/translator feedback?)

Note: It took a minute of fiddling to get the namespaced example to not get entangled with the markdown in this comment--all in service of getting the underscore on the right side.

Reply:
I think the proposed solution solves the 99% case, and trying to solve the remaining 1% will be tricky. Moreover, the people who really have to deal with messages are the translators, and they will need tooling to effectively do their jobs at scale — and that tooling has *far* more freedom to deal with bidi issues than we have in plain text. So I don't think we need to go any further down this path.

Reply:
My position on this is that if we're going to accommodate RTL literals and identifiers, then we should endeavour to do so completely, and AFAIK this is the only remaining place where we'd need bidi controls in order to do that.

> users should be careful about using mixed direction for non-literal values

Tbf, `_` has neutral direction and it is relatively common to use it as a prefix for identifiers.

Reply:
> Tbf, _ has neutral direction and it is relatively common to use it as a prefix for identifiers.

Yes, I know. So do all of the sigils (with some of them being enclosing punctuation). 

I guess my argument would be that the sigil is, from a user perspective, "part of the name", e.g. `$foo`, not `$` with `foo` being separate. In an RTL display context, the sigil, being a neutral, will naturally reverse sides to be at the start, (e.g. `
⁧$_مصر⁩
`). For us to maintain an LTR bias while making _only_ the name token isolated is a lot more work--for tools and users.

The counterargument would be that `identifier` and `name` are their own things (and many are generated from the user's environment). The various sigils and such in MF2 are thus "not part of the name" and should be treated separately. Allowing users to insert these characters is not the same as obligating them to use them and I could see us making the affordance.

In the screenshot below, the display is set to RTL. The placeholders have an LRI/PDI inside the `{`/`}`. 
- The first line, lacking any other bidi controls, is illegible. 
- The second line uses RLI/PDI around each token. This presents sigils on the right, which an RTL speaker would understand, but the tokens are in LTR order. I don't specifically object to this presentation.
- The third line uses LRI/PDI around each token and approximates what I'm suggesting above.
- The fourth line uses the markup you're suggesting, which is RLI/PDI around each `name` and LRI/PDI around the sigil+identifier.


![image](https://github.com/user-attachments/assets/ca005661-4010-4f0c-8412-2245b4b445b0)


Here's the data used to produce the screenshot. Note that I do use an U+200E next to the `:` separator.

```
Example: {\u2066$_\u06351\u0636\u0637 :_\u06352\u0636\u0637:_\u06353\u0636\u0637 _\u06354\u0636\u0637=_\u06355\u0636\u0637\u2069}<br>
Example RLI: {\u2066\u2067$_\u06351\u0636\u0637\u2069 \u2067:_\u06352\u0636\u0637:_\u06353\u0636\u0637\u2069 \u2067_\u06354\u0636\u0637\u2069=\u2067_\u06355\u0636\u0637\u2069\u2069}<br>
Example LRI: {\u2066\u2066$_\u06351\u0636\u0637\u2069 \u2066:_\u06352\u0636\u0637\u200e:_\u06353\u0636\u0637\u2069 \u2066_\u06354\u0636\u0637\u2069=\u2066_\u06355\u0636\u0637\u2069\u2069}<br>
Example EAO: {\u2066\u2066$\u2067_\u06351\u0636\u0637\u2069\u2069 \u2066:\u2067_\u06352\u0636\u0637\u2069\u200e:\u2067_\u06353\u0636\u0637\u2069\u2069 \u2067_\u06354\u0636\u0637\u2069=\u2067_\u06355\u0636\u0637\u2069\u2069}<br>
```

Reply:
> The various sigils and such in MF2 are thus "not part of the name" and should be treated separately.

That would be my position. In particular for variables, AFAIK all current implementations would require parameters like `{ foo: 42 }` in order to resolve a `$foo`, and not e.g. ` { $foo: 42 }`.

In case it matters, here's a slightly shorter way to get the same results as in the last example (The LR isolation around the sigil+identifier isn't required as the placeholder already isolates its contents, and the LRM next to the `:` has no effect):

```
Example EAO: {\u2066$\u2067_\u06351\u0636\u0637\u2069 :\u2067_\u06352\u0636\u0637\u2069:\u2067_\u06353\u0636\u0637\u2069 \u2067_\u06354\u0636\u0637\u2069=\u2067_\u06355\u0636\u0637\u2069\u2069}
```

Reply:
I think something got damaged somewhere in the several copy/paste operations.

To make things readable I replaced:
```
\u0635 A
\u0636 B
\u0637 C

\u2066 [LRI]
\u2069 [PDI]
\u200E [LRM]
\u2067 [RLI]
```

And the resulting strings above are:
```
Example: {[LRI]$_A1BC :_A2BC:_A3BC _A4BC=_A5BC[PDI]}<br>
Example RLI: {[LRI][RLI]$_A1BC[PDI] [RLI]:_A2BC:_A3BC[PDI] [RLI]_A4BC[PDI]=[RLI]_A5BC[PDI][PDI]}<br>
Example LRI: {[LRI][LRI]$_A1BC[PDI] [LRI]:_A2BC[LRM]:_A3BC[PDI] [LRI]_A4BC[PDI]=[LRI]_A5BC[PDI][PDI]}<br>
Example EAO: {[LRI][LRI]$[RLI]_A1BC[PDI][PDI] [LRI]:[RLI]_A2BC[PDI][LRM]:[RLI]_A3BC[PDI][PDI] [RLI]_A4BC[PDI]=[RLI]_A5BC[PDI][PDI]}<br>

Example EAO: {[LRI]$[RLI]_A1BC[PDI] :[RLI]_A2BC[PDI]:[RLI]_A3BC[PDI] [RLI]_A4BC[PDI]=[RLI]_A5BC[PDI][PDI]}
```

Taking out the bidi control characters we are left with
```
{$_A1BC :_A2BC:_A3BC _A4BC=_A5BC}
```

That is not valid MF2 syntax.
So I don't know what this is trying to fix.

---

But yesterday I played a bit and put together a web page that one can use to interactively play with this.

https://mihai-nita.net/tmp/mf2bidi.html

And I argue that:
- isolates are enough (FSI, LRI, RLI, PDI)
- there is no need for any control character between `$` and the name proper, even when there is an `_` there.

Reply:
> That is not valid MF2 syntax.

Um... how is it not valid?

`$_A1BC` is a valid operand.
`:_A2BC` is a valid namespace. `:_A3BC` is a valid function name.
`_A4BC` is a valid option name.
`_A5BC` is a valid unquoted literal.

It's even a valid message, in that a simple message might consist solely of a placeholder.

What am I missing?